src/nvim/api/mark_extended.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Same function but names are not limited to one char
+ * There is also no ex_command, just viml functions
+ */
+
+#include "nvim/vim.h"
+#include "nvim/globals.h"      // FOR_ALL_BUFFERS
+#include "nvim/mark.h"         // SET_FMARK
+#include "nvim/memory.h"
+#include "nvim/map.h"          // pmap ...
+#include "nvim/lib/kbtree.h"   // kbitr ...
+#include "nvim/api/mark_extended.h"
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "api/extmark.c.generated.h"
+#endif
+
+#define FOR_ALL_EXTMARKS(buf) \
+  kbitr_t itr; \
+  ExtendedMark *extmark; \
+  for (;kb_itr_valid(&itr); kb_itr_next(ExtendedMark, buf->b_extmarks_tree, &itr)){
+
+/* Create or update an extmark, */
+int extmark_set(buf_T *buf, cstr_t *name, pos_T *pos)
```

Comment:
I did't say wrappers must be used - just that some operations are common enough to justify it, ie a wrapper `kh_putval(type, somestruct->map, key, val)` instead of `kh_val(somestruct->map, kh_put(type, somestruct->map, key, &ret) = val`. This can just be inline functions or macros, no need for a separate compilation unit `map.c`.

> My best guess is that this has something to do with CPU cache, three allocations are definitely not going to be faster then one.

The vectors are going to be used many times for each (re)allocation (though in principle a struct-of-arrays could also be allocated using a single alloc just fine, if one just places the most aligned type first) and as flags and keys will be accessed a lot more than values in lookup it is often beneficial to keep them close together. (though an another option is to keep the keys and values array compact, and have the sparse hash table just have indicies to them, like [pypy](https://morepypy.blogspot.se/2015/01/faster-more-memory-efficient-and-more.html):s implementation, though the approaches could also be combined )